### PR TITLE
OAuth 2.0 retry

### PR DESCRIPTION
**Pull request by alex-schneider**

### CODE REVIEW

File: server/testdata/endpoints/04_couper.hcl [hunk 1]
```diff
@@ -8,13 +8,14 @@ server "api" {
...
-		  path   = "/resource"
+          path   = "/resource"
 
-		  oauth2 {
+          oauth2 {
             token_endpoint = "${request.headers.x-token-endpoint}/oauth2"
             client_id      = "user"
             client_secret  = "pass"
             grant_type     = "client_credentials"
+            retries        = 2
```
Comment (malud) [inline server/testdata/endpoints/04_couper.hcl:18]: Testcase for `0` is missing.

File: handler/transport/oauth2.go [hunk 2]
```diff
@@ -40,6 +40,14 @@ func NewOAuth2(config *config.OAuth2, memStore *cache.MemoryStore,
 		return nil, fmt.Errorf("the grant_type has to be set to 'client_credentials'")
 	}
 
+	if config.Retries == nil {
```
Comment (malud) [inline handler/transport/oauth2.go:43]: Using `*uint8` would give the responsibility for negative numbers to the hcl decoding.
Additionally its range is capped by type: 0 - 255 which should be enough retries.
If this is `nil` the default can be set... on config level. The specific handling here would be obsolete.

File: server/testdata/endpoints/04_couper.hcl (hunk 1 above)
Comment (malud) [reply server/testdata/endpoints/04_couper.hcl:18]: And if the default of `1` is set and executed if the attribute is missing.

File: server/http_endpoints_test.go [hunk 3]
```diff
@@ -263,6 +263,7 @@ func TestEndpoints_UpstreamBasicAuthAndXFF(t *testing.T) {
 func TestEndpoints_OAuth2(t *testing.T) {
```
Comment (malud) [inline server/http_endpoints_test.go:263]: As mentioned below some cases are missing. Replacing an existing test, which seems to test the default behaviour, is not helpful. Add additional cases to test the default, disabled and >default .

**Review by malud — CHANGES_REQUESTED** [review]

File: handler/transport/oauth2.go (hunk 2 above)
Comment (alex-schneider) [reply handler/transport/oauth2.go]: Done

File: server/testdata/endpoints/04_couper.hcl (hunk 1 above)
Comment (alex-schneider) [reply server/testdata/endpoints/04_couper.hcl]: Done

File: server/http_endpoints_test.go (hunk 3 above)
Comment (alex-schneider) [reply server/http_endpoints_test.go]: Done

**Review by malud — APPROVED** [review]

**Pull request merged**